File: folio_app/services/comments.py
from __future__ import annotations

from collections import Counter
from dataclasses import dataclass
from datetime import datetime, timezone
import logging
from typing import Any

import streamlit as st

from folio_app.services.supabase_client import get_supabase_client


logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class CommentResult:
    ok: bool
    message: str
    comment: dict[str, Any] | None = None
    comment_id: str | None = None


def list_project_comments(project_id: str) -> list[dict[str, Any]]:
    client = get_supabase_client()
    if client is None:
        return []

    try:
        response = (
            client.table("comments")
            .select("*")
            .eq("project_id", project_id)
            .order("created_at", desc=False)
            .execute()
        )
        return _attach_comment_authors(response.data or [])
    except Exception:
        logger.exception("Failed to list project comments")
        return []


def create_comment(project_id: str, author_id: str, body: str, parent_id: str | None = None) -> CommentResult:
    from folio_app.services.auth import ensure_authenticated_session

    normalized_body = body.strip()
    if not normalized_body:
        return CommentResult(False, "댓글 내용을 입력하세요.")
    if len(normalized_body) > 1000:
        return CommentResult(False, "댓글은 1,000자 이내로 입력하세요.")

    auth_result = ensure_authenticated_session()
    if not auth_result.ok:
        return CommentResult(False, auth_result.message)

    client = get_supabase_client()
    if client is None:
        return CommentResult(False, "Supabase 환경 변수가 설정되지 않았습니다.")

    if parent_id and not _can_reply_to_comment(project_id, parent_id):
        return CommentResult(False, "답글을 남길 댓글을 확인할 수 없습니다.")

    payload = {
        "project_id": project_id,
        "author_id": author_id,
        "body": normalized_body,
        "parent_id": parent_id,
        "depth": 1 if parent_id else 0,
    }

    try:
        response = (
            client.table("comments")
            .insert(payload)
            .select("*")
            .execute()
        )
        if not response.data:
            return CommentResult(False, "댓글 등록에 실패했습니다.")
    except Exception:
        logger.exception("Failed to create project comment")
        return CommentResult(False, "댓글 등록에 실패했습니다. 잠시 후 다시 시도하세요.")

    clear_comment_caches()
    created = _attach_comment_authors(response.data or [])
    created_comment = created[0] if created else response.data[0]
    _create_comment_notification(project_id, created_comment, author_id)
    return CommentResult(True, "댓글이 등록되었습니다.", comment=created_comment)


def delete_comment(comment_id: str, author_id: str) -> CommentResult:
    from folio_app.services.auth import ensure_authenticated_session

    auth_result = ensure_authenticated_session()
    if not auth_result.ok:
        return CommentResult(False, auth_result.message)

    client = get_supabase_client()
    if client is None:
        return CommentResult(False, "Supabase 환경 변수가 설정되지 않았습니다.")

    try:
        response = (
            client.table("comments")
            .delete()
            .eq("id", comment_id)
            .eq("author_id", author_id)
            .execute()
        )
        if not response.data:
            return CommentResult(False, "삭제할 댓글을 찾을 수 없습니다.")
    except Exception:
        logger.exception("Failed to delete project comment")
        return CommentResult(False, "댓글 삭제에 실패했습니다. 잠시 후 다시 시도하세요.")

    clear_comment_caches()
    return CommentResult(True, "댓글이 삭제되었습니다.", comment_id=comment_id)


def build_comment_tree(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    by_id = {row["id"]: {**row, "children": []} for row in rows}
    roots: list[dict[str, Any]] = []

    for row in rows:
        node = by_id[row["id"]]
        parent_id = row.get("parent_id")
        parent = by_id.get(parent_id) if parent_id else None
        if parent and not parent.get("parent_id") and (parent.get("depth") or 0) == 0:
            parent["children"].append(node)
        elif parent and parent.get("parent_id") in by_id:
            by_id[parent["parent_id"]]["children"].append(node)
        else:
            roots.append(node)

    return roots


def count_comments_by_project(project_ids: list[str]) -> dict[str, int]:
    return _fetch_comment_counts(tuple(sorted(set(project_ids))))


def latest_comment_at_by_project(project_ids: list[str]) -> dict[str, str]:
    return _fetch_latest_comment_times(tuple(sorted(set(project_ids))))


def annotate_unread_comment_status(projects: list[dict[str, Any]], user_id: str) -> list[dict[str, Any]]:
    unread_project_ids = get_unread_comment_project_ids(projects, user_id)
    for project in projects:
        project["has_unread_comments"] = project.get("id") in unread_project_ids
    return projects


def get_unread_comment_project_ids(projects: list[dict[str, Any]], user_id: str) -> set[str]:
    client = get_supabase_client()
    project_ids = [project["id"] for project in projects if project.get("id")]
    if client is None or not project_ids or not user_id:
        return set()

    try:
        comments_response = (
            client.table("comments")
            .select("project_id, author_id, created_at")
            .in_("project_id", project_ids)
            .neq("author_id", user_id)
            .execute()
        )
        reads_response = (
            client.table("project_comment_reads")
            .select("project_id, last_read_at")
            .eq("user_id", user_id)
            .in_("project_id", project_ids)
            .execute()
        )
    except Exception:
        logger.exception("Failed to load unread comment state")
        st.session_state["portfolio_unread_comment_error"] = (
            "새 댓글 표시 상태를 확인하지 못했습니다. Supabase 스키마 적용 여부를 확인하세요."
        )
        return set()

    latest_external_comments: dict[str, datetime] = {}
    for comment in comments_response.data or []:
        project_id = comment.get("project_id")
        created_at = _parse_timestamp(comment.get("created_at"))
        if not project_id or created_at is None:
            continue
        current = latest_external_comments.get(project_id)
        if current is None or created_at > current:
            latest_external_comments[project_id] = created_at

    reads_by_project = {
        read["project_id"]: _parse_timestamp(read.get("last_read_at"))
        for read in reads_response.data or []
        if read.get("project_id")
    }

    unread_project_ids: set[str] = set()
    for project_id, latest_comment_at in latest_external_comments.items():
        last_read_at = reads_by_project.get(project_id)
        if last_read_at is None or latest_comment_at > last_read_at:
            unread_project_ids.add(project_id)
    return unread_project_ids


def mark_project_comments_read(project_id: str, user_id: str) -> bool:
    from folio_app.services.auth import ensure_authenticated_session

    if not project_id or not user_id:
        return False

    auth_result = ensure_authenticated_session()
    if not auth_result.ok:
        return False

    client = get_supabase_client()
    if client is None:
        return False

    now = datetime.now(timezone.utc).isoformat()
    try:
        (
            client.table("project_comment_reads")
            .upsert(
                {
                    "project_id": project_id,
                    "user_id": user_id,
                    "last_read_at": now,
                    "updated_at": now,
                },
                on_conflict="project_id,user_id",
            )
            .execute()
        )
    except Exception:
        logger.exception("Failed to mark project comments as read")
        return False
    return True


@st.cache_data(ttl=15, show_spinner=False)
def _fetch_comment_counts(project_ids: tuple[str, ...]) -> dict[str, int]:
    client = get_supabase_client()
    if not project_ids:
        return {}
    if client is None:
        return {}

    try:
        response = (
            client.table("comments")
            .select("project_id")
            .in_("project_id", list(project_ids))
            .execute()
        )
    except Exception:
        logger.exception("Failed to count project comments")
        return {}

    counter: Counter[str] = Counter()
    for comment in response.data or []:
        project_id = comment.get("project_id")
        if project_id:
            counter[project_id] += 1
    return dict(counter)


@st.cache_data(ttl=15, show_spinner=False)
def _fetch_latest_comment_times(project_ids: tuple[str, ...]) -> dict[str, str]:
    client = get_supabase_client()
    if not project_ids:
        return {}
    if client is None:
        return {}

    try:
        response = (
            client.table("comments")
            .select("project_id, created_at")
            .in_("project_id", list(project_ids))
            .execute()
        )
    except Exception:
        logger.exception("Failed to load latest project comment times")
        return {}

    latest_by_project: dict[str, datetime] = {}
    latest_raw_by_project: dict[str, str] = {}
    for comment in response.data or []:
        project_id = comment.get("project_id")
        raw_created_at = comment.get("created_at")
        created_at = _parse_timestamp(raw_created_at)
        if not project_id or created_at is None:
            continue
        current = latest_by_project.get(project_id)
        if current is None or created_at > current:
            latest_by_project[project_id] = created_at
            latest_raw_by_project[project_id] = raw_created_at
    return latest_raw_by_project


def clear_comment_caches() -> None:
    _fetch_comment_counts.clear()
    _fetch_latest_comment_times.clear()


def _can_reply_to_comment(project_id: str, parent_id: str) -> bool:
    client = get_supabase_client()
    if client is None:
        return False

    try:
        response = (
            client.table("comments")
            .select("id, project_id, parent_id, depth")
            .eq("id", parent_id)
            .limit(1)
            .execute()
        )
    except Exception:
        logger.exception("Failed to validate reply target")
        return False

    parent = (response.data or [None])[0]
    if not parent:
        return False
    return (
        parent.get("project_id") == project_id
        and parent.get("parent_id") is None
        and (parent.get("depth") or 0) == 0
    )


def _attach_comment_authors(comments: list[dict[str, Any]]) -> list[dict[str, Any]]:
    if not comments:
        return []

    client = get_supabase_client()
    if client is None:
        return comments

    author_ids = sorted({comment["author_id"] for comment in comments if comment.get("author_id")})
    profiles_by_id: dict[str, dict[str, Any]] = {}
    if author_ids:
        try:
            response = (
                client.table("public_profiles")
                .select("id, name")
                .in_("id", author_ids)
                .execute()
            )
            profiles_by_id = {profile["id"]: profile for profile in response.data or []}
        except Exception:
            logger.exception("Failed to attach comment authors")

    return [
        {
            **comment,
            "author": profiles_by_id.get(comment.get("author_id"), {}),
        }
        for comment in comments
    ]


def _parse_timestamp(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        return parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _create_comment_notification(project_id: str, comment: dict[str, Any], author_id: str) -> None:
    try:
        from folio_app.services.notifications import create_project_comment_notification

        create_project_comment_notification(project_id, comment, author_id)
    except Exception:
        logger.exception("Failed to dispatch comment notification")

File: svelte_app/src/lib/comments.ts
import { currentSession } from '$lib/auth';
import { getSupabaseClient } from '$lib/supabase';
import type { ProjectComment } from '$lib/types';

type CommentRow = Omit<ProjectComment, 'author' | 'children'>;

type CommentMutationOptions = {
	projectAuthorId?: string | null;
	projectTitle?: string | null;
};

export async function listProjectComments(projectId: string) {
	const supabase = getSupabaseClient();
	if (!supabase) {
		return {
			comments: [],
			error: 'Supabase 환경 변수가 설정되지 않았습니다.'
		};
	}

	const { data, error } = await supabase
		.from('comments')
		.select('id,project_id,author_id,parent_id,body,depth,is_deleted,created_at')
		.eq('project_id', projectId)
		.order('created_at', { ascending: true });

	if (error) {
		return {
			comments: [],
			error: '댓글을 불러오지 못했습니다.'
		};
	}

	const comments = await attachCommentAuthors((Array.isArray(data) ? data : []).map(normalizeComment));
	return {
		comments: buildCommentTree(comments),
		error: ''
	};
}

export async function createProjectComment(projectId: string, body: string, options: CommentMutationOptions = {}) {
	return createComment(projectId, body, null, options);
}

export async function createProjectReply(projectId: string, parentId: string, body: string, options: CommentMutationOptions = {}) {
	return createComment(projectId, body, parentId, options);
}

export async function deleteProjectComment(commentId: string) {
	const supabase = getSupabaseClient();
	const session = await currentSession();
	if (!supabase || !session) {
		return { ok: false, message: '로그인 후 댓글을 삭제할 수 있습니다.' };
	}

	const { error } = await supabase.from('comments').delete().eq('id', commentId).eq('author_id', session.user.id);
	if (error) {
		return { ok: false, message: '댓글 삭제에 실패했습니다. 잠시 후 다시 시도하세요.' };
	}

	return { ok: true, message: '댓글이 삭제되었습니다.' };
}

export async function currentCommentUserId() {
	const session = await currentSession();
	return session?.user.id ?? null;
}

export async function isCommentAuthenticated() {
	return Boolean(await currentSession());
}

export async function markProjectCommentsSeen(projectId: string, projectAuthorId?: string | null) {
	const supabase = getSupabaseClient();
	const session = await currentSession();
	if (!supabase || !session || !projectAuthorId || session.user.id !== projectAuthorId) {
		return false;
	}

	const now = new Date().toISOString();
	await supabase
		.from('project_comment_reads')
		.upsert(
			{
				project_id: projectId,
				user_id: session.user.id,
				last_read_at: now,
				updated_at: now
			},
			{ onConflict: 'project_id,user_id' }
		);

	await supabase
		.from('notifications')
		.update({ is_read: true, read_at: now })
		.eq('user_id', session.user.id)
		.eq('project_id', projectId)
		.eq('type', 'project_comment')
		.eq('is_read', false);

	return true;
}

async function createComment(projectId: string, body: string, parentId: string | null, options: CommentMutationOptions) {
	const supabase = getSupabaseClient();
	const session = await currentSession();
	if (!supabase || !session) {
		return { ok: false, message: '로그인 후 댓글을 작성할 수 있습니다.' };
	}

	const normalizedBody = body.trim();
	if (!normalizedBody) {
		return { ok: false, message: '댓글 내용을 입력하세요.' };
	}
	if (normalizedBody.length > 1000) {
		return { ok: false, message: '댓글은 1,000자 이내로 입력하세요.' };
	}

	const { data, error } = await supabase
		.from('comments')
		.insert({
			project_id: projectId,
			author_id: session.user.id,
			body: normalizedBody,
			parent_id: parentId,
			depth: parentId ? 1 : 0
		})
		.select('id,project_id,author_id,parent_id,body,depth,is_deleted,created_at')
		.single();

	if (error) {
		return { ok: false, message: '댓글 등록에 실패했습니다. 잠시 후 다시 시도하세요.' };
	}

	await createCommentNotification(projectId, normalizeComment(data), session.user.id, options);
	void sendCommentEmailNotification(String(data.id ?? ''));
	return { ok: true, message: '댓글이 등록되었습니다.' };
}

async function createCommentNotification(
	projectId: string,
	comment: ProjectComment,
	actorId: string,
	options: CommentMutationOptions
) {
	const supabase = getSupabaseClient();
	if (!supabase || !comment.id || !options.projectAuthorId || options.projectAuthorId === actorId) {
		return;
	}

	const { error } = await supabase.from('notifications').insert({
		user_id: options.projectAuthorId,
		actor_id: actorId,
		project_id: projectId,
		comment_id: comment.id,
		type: 'project_comment',
		title: `${options.projectTitle || '프로젝트'}에 새 댓글이 남겨졌습니다.`,
		body: comment.body
	});

	if (error && !isUniqueViolation(error)) {
		console.warn('Failed to create comment notification', error);
	}
}

async function sendCommentEmailNotification(commentId: string) {
	const session = await currentSession();
	if (!session || !commentId) {
		return;
	}
	const response = await fetch(`/api/comments/${commentId}/email-notification`, {
		method: 'POST',
		headers: {
			Authorization: `Bearer ${session.access_token}`
		}
	});
	if (!response.ok && response.status !== 202) {
		console.warn('Failed to request comment email notification');
	}
}

async function attachCommentAuthors(comments: ProjectComment[]) {
	const supabase = getSupabaseClient();
	if (!supabase || comments.length === 0) {
		return comments;
	}

	const authorIds = [...new Set(comments.map((comment) => comment.author_id).filter(Boolean))];
	const { data } = await supabase.from('public_profiles').select('id,name').in('id', authorIds);
	const profileById = new Map(
		(Array.isArray(data) ? data : []).map((profile) => {
			const record = asRecord(profile);
			return [String(record.id), { id: nullableString(record.id) ?? undefined, name: nullableString(record.name) ?? undefined }];
		})
	);

	return comments.map((comment) => ({
		...comment,
		author: profileById.get(comment.author_id) ?? {}
	}));
}

function buildCommentTree(comments: ProjectComment[]) {
	const byId = new Map(comments.map((comment) => [comment.id, { ...comment, children: [] as ProjectComment[] }]));
	const roots: ProjectComment[] = [];

	for (const comment of byId.values()) {
		if (comment.parent_id && byId.has(comment.parent_id)) {
			const parent = byId.get(comment.parent_id)!;
			parent.children.push(comment);
		} else if (comment.parent_id) {
			roots.push({ ...comment, parent_id: null, depth: 0 });
		} else {
			roots.push(comment);
		}
	}

	return roots;
}

function normalizeComment(value: unknown): ProjectComment {
	const payload = asRecord(value) as Partial<CommentRow>;
	return {
		id: String(payload.id ?? ''),
		project_id: String(payload.project_id ?? ''),
		author_id: String(payload.author_id ?? ''),
		parent_id: nullableString(payload.parent_id),
		body: String(payload.body ?? ''),
		depth: Number(payload.depth ?? 0) === 1 ? 1 : 0,
		is_deleted: Boolean(payload.is_deleted ?? false),
		created_at: String(payload.created_at ?? ''),
		author: {},
		children: []
	};
}

function asRecord(value: unknown): Record<string, unknown> {
	return value && typeof value === 'object' ? (value as Record<string, unknown>) : {};
}

function nullableString(value: unknown) {
	const text = String(value ?? '').trim();
	return text || null;
}

function isUniqueViolation(error: unknown) {
	const record = asRecord(error);
	const code = String(record.code ?? '');
	const message = String(record.message ?? '').toLowerCase();
	return code === '23505' || message.includes('duplicate key') || message.includes('unique constraint');
}
